File: backend/internal/live/incident.go
package live

// Incident context generation (spec §17, §18), ported from the
// generateIncidentContextAction flow: fetch the two panels (cached), peek the
// cross-panel caches, fold in SQLite history, render.

import (
	"context"
	"time"

	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/analysis"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/cache"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/service"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/types"
)

func (p *Provider) IncidentContext(ctx context.Context) (types.IncidentContextResult, error) {
	now := p.Now()
	win := service.ResolveWindow(nil, now.UnixMilli())

	var metrics *types.MetricsPanel
	if m, err := p.MetricsPanel(ctx, win); err == nil {
		metrics = &m
	}
	var kubePanel *types.KubePanel
	if k, err := p.KubePanel(ctx); err == nil {
		kubePanel = &k
	}
	fingerprints, _ := cache.Peek[[]types.FingerprintEntry]("panel:fingerprints")
	verifications, _ := cache.Peek[[]types.VerificationResult]("panel:verifications")

	var logs, prevLogs *analysis.LogsRef
	if l, ok := cache.Peek[analysis.LogsRef]("panel:lastlogs"); ok {
		logs = &l
	}
	if l, ok := cache.Peek[analysis.LogsRef]("panel:lastprevlogs"); ok {
		prevLogs = &l
	}

	parts := analysis.IncidentParts{
		Fingerprints:  fingerprints,
		Logs:          logs,
		PreviousLogs:  prevLogs,
		Verifications: verifications,
	}
	if metrics != nil {
		parts.Metrics = metrics.Metrics
		parts.HttpSummary = metrics.HttpSummary
		parts.Anomalies = metrics.Anomalies
		parts.Correlations = metrics.Correlations
		parts.Timeline = metrics.Timeline
	}
	parts.Kube = kubePanel

	// History unavailable — snapshot proceeds with live data only.
	if rows, err := p.Store.ListWafHistoryRows(); err == nil {
		for _, h := range rows {
			parts.WafHistory = append(parts.WafHistory, analysis.WafHistoryEntry{
				Ts:       time.UnixMilli(h.Ts).UTC().Format(time.RFC3339Nano),
				RuleName: h.RuleName, Action: h.Action, Status: h.Status, Detail: h.Detail,
			})
		}
	}
	if rows, err := p.Store.ListDeployHistory(); err == nil {
		for _, d := range rows {
			parts.DeployHistory = append(parts.DeployHistory, analysis.DeployHistoryEntry{
				Ts:     time.UnixMilli(d.Ts).UTC().Format(time.RFC3339Nano),
				Target: d.Namespace + "/" + d.Name, Change: d.Change, Verdict: d.Verdict,
			})
		}
	}

	snapshot := analysis.BuildSnapshot(parts, now)
	json := analysis.ToJson(snapshot)
	// Persisting the snapshot is best-effort.
	_ = p.Store.SaveIncidentSnapshot(json, now.UnixMilli())

	return types.IncidentContextResult{
		Markdown:    analysis.ToMarkdown(snapshot),
		Json:        json,
		QPrompt:     analysis.ToQPrompt(snapshot),
		GeneratedAt: snapshot.Timestamp,
	}, nil
}

File: backend/live/incident.ts
// Incident context generation (spec §17, §18): fetch the two panels (cached),
// peek the cross-panel caches, fold in SQLite history, render.

import {
  buildSnapshot,
  toJson,
  toMarkdown,
  toQPrompt,
  type DeployHistoryEntry,
  type IncidentParts,
  type LogsRef,
  type WafHistoryEntry,
} from "../analysis/incident.ts";
import { peek } from "../cache/cache.ts";
import { resolveWindow } from "../service/window.ts";
import type {
  FingerprintEntry,
  IncidentContextResult,
  KubePanel,
  VerificationResult,
} from "../../src/lib/types.ts";
import type { MetricsPanelFull } from "../types/types.ts";
import type { LiveProvider } from "./live.ts";

export async function incidentContext(p: LiveProvider): Promise<IncidentContextResult> {
  const nowMs = p.now();
  const win = resolveWindow(null, nowMs);

  let metrics: MetricsPanelFull | null = null;
  try {
    metrics = await p.metricsPanel(win);
  } catch {
    // A snapshot with no metrics is still worth producing.
  }
  let kubePanel: KubePanel | null = null;
  try {
    kubePanel = await p.kubePanel();
  } catch {
    // ditto
  }

  const parts: IncidentParts = {
    metrics: metrics?.metrics ?? [],
    httpSummary: metrics?.httpSummary ?? null,
    kube: kubePanel,
    anomalies: metrics?.anomalies ?? [],
    correlations: metrics?.correlations ?? [],
    timeline: metrics?.timeline ?? [],
    fingerprints: peek<FingerprintEntry[]>("panel:fingerprints") ?? [],
    logs: peek<LogsRef>("panel:lastlogs") ?? null,
    previousLogs: peek<LogsRef>("panel:lastprevlogs") ?? null,
    wafHistory: [],
    deployHistory: [],
    verifications: peek<VerificationResult[]>("panel:verifications") ?? [],
  };

  // History unavailable — the snapshot proceeds with live data only.
  try {
    parts.wafHistory = p.store.listWafHistoryRows().map(
      (h): WafHistoryEntry => ({
        ts: new Date(h.ts).toISOString(),
        ruleName: h.ruleName,
        action: h.action,
        status: h.status,
        detail: h.detail,
      }),
    );
  } catch {
    // ignored
  }
  try {
    parts.deployHistory = p.store.listDeployHistory().map(
      (d): DeployHistoryEntry => ({
        ts: new Date(d.ts).toISOString(),
        target: `${d.namespace}/${d.name}`,
        change: d.change,
        verdict: d.verdict,
      }),
    );
  } catch {
    // ignored
  }

  const snapshot = buildSnapshot(parts, new Date(nowMs));
  const json = toJson(snapshot);
  // Persisting the snapshot is best-effort.
  try {
    p.store.saveIncidentSnapshot(json, nowMs);
  } catch {
    // ignored
  }

  return {
    markdown: toMarkdown(snapshot),
    json,
    qPrompt: toQPrompt(snapshot),
    generatedAt: snapshot.timestamp,
  };
}
